update: use java.util.BitSet on js platform

i10416 opened the pull request:
close #381 
scala.js supports java.util.BitSet since 1.9.0

johnynek commented:
I don't think we can merge this and #376.

Also the tests will be very slow if we don't have the switch for isScalaJs.

Can scala-native handle Java util Bitset?

Finally, this will be binary incompatible.

johnynek commented:
But *thanks* for the PR!!

Hopefully we can merge some version.

i10416 commented:
Thank you for review. I will tweak what you pointed out.

> I don't think we can merge this and #376

Yes, I think too.
I planed to share source between jvm/js first, then tweak #376 

> Can scala-native handle Java util Bitset?
 
AFAIK, scala native does not support java.util.BitSet :(

i10416 commented:
> Also the tests will be very slow if we don't have the switch for isScalaJs.

I think it would be better to switch scala.js/jvm tests based on build settings (e.g. generate util object as [sbt-buildinfo](https://github.com/sbt/sbt-buildinfo) do ) rather than `BitSetUtil` fields. What do you think?

i10416 commented:
Ah, I found https://github.com/portable-scala/sbt-crossproject/issues/17. IIUC, we need to copy-paste `BitSetUtil` between jvm and js to add native platform impl later?

armanbilge commented:
@i10416 I'm actually about to submit a PR to fix https://github.com/portable-scala/sbt-crossproject/issues/17 :)

> AFAIK, scala native does not support java.util.BitSet :(

That's true, but now that Scala.js does you can probably just copy-paste the implementation and tests. That's how many things get ported to native :)

> I think it would be better to switch scala.js/jvm tests based on build settings

I think if you have a dependency on cats-laws in the tests you can just use this:

https://github.com/typelevel/cats/blob/b307688c391a47dc6dd82d42c15edf273bf2471b/kernel-laws/jvm/src/main/scala/cats/platform/Platform.scala

armanbilge commented:
- https://github.com/scala-native/scala-native/pull/2565

i10416 commented:
> tests will be very slow if we don't have the switch for isScalaJs.

Now that test iteration is switched based on js/jvm, what to do are

1. merge #384 in future release(or is there a nice trick to keep bincompat?)
2. wait for https://github.com/scala-native/scala-native/pull/2565 to be merged/released(thx @armanbilge )
3. then tweak and finish #376.

johnynek commented:
I think it is reasonable to release `0.4.0` and break this minor binary compatibility (isScalaJs, isScalaJvm)...

Those really should have been made `private [cats.parse]` anyway...

armanbilge commented:
FWIW I'd appreciate not bumping and preserving binary-compatibility here instead. It makes downstream cross-builds slightly more complicated when the different platforms are pinned on different versions of the same dependency.

Apologies if I am missing something, but why is it so hard to preserve bincompat here? Or is it just undesirable?

johnynek commented:
@armanbilge I'm confused by what you mean.

The only real break to binary compatibility is the symbols `isScalaJvm` and `isScalaJs`. I bet no one is actually using those. So, any real dependency won't be broken.

The other errors are for items inside the `private object Impl` which no one can see without doing some weird reflection hacks, which we don't promise to support.

Do you think there are downstream users that care about the first two?

Or is it something about using scalajs 1.9.0?

armanbilge commented:
FTR I completely agree that its unlikely that removing those will break any downstreams. 

However, it's the version bump that got my attention: it will force downstreams to also bump their versions to take on this update.

johnynek commented:
okay, fair point.

Let's add the isScalaJvm, isScalaJs back even though it is inelegant. We can add exclusions for the changes inside `private object Impl`.

i10416 commented:
Thank you for reviewing. I will put isScalaJvm and isScalaJs back.

i10416 commented on core/shared/src/main/scala/cats/parse/BitSet.scala:
```diff
@@ -22,12 +22,12 @@
 object BitSetUtil {
   type Tpe = BitSet
-
+  @deprecated("", "0.3.7")
```
I changed  here to `isScalaJVM =false , isScalaJs = false` because I think combination of `isScalaJVM =true, isScalaJs = false`, `isScalaJVM =false, isScalaJs = true` or `isScalaJVM =true, isScalaJs = true` does not make sense.

i10416 replied:
> instead we could move the BitSet code to ...

Yeah, actually, I thought my idea was not nice too😓

That seems better solution! Thanks!

i10416 commented:
Hmmm, can't I write exclude rules like `cats.parse.Parser#Impl.*` ?

i10416 commented on core/shared/src/test/scala/cats/parse/LocationMapTest.scala:
```diff
@@ -26,7 +26,7 @@ import Prop.forAll
 class LocationMapTest extends munit.ScalaCheckSuite {
 
-  val tests: Int = if (BitSetUtil.isScalaJs) 50 else 20000
+  val tests: Int = 20000
```
Oh, I forgot replacing here with TestUtils.iteration😓

i10416 commented:
I removed deprecated annotations, which I added just in case though I didn't think someone uses `isScalaJs` and `isScalaJvm` api, because I assume you don't like warning on the ground that you regard warning as fatal when checking source before marge (while I felt it makes sense to show deprecated warning).

i10416 commented:
There remain some mima warnings such as `method bitSetFor(Array[Char])scala.collection.mutable.BitSet in object cats.parse.BitSetUtil has a different result type in current version, where it is java.util.BitSet rather than scala.collection.mutable.BitSet filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("cats.parse.BitSetUtil.bitSetFor")`. 

Should I suppress them too for now?  If we would suppress them, I think we would need to remove the exclude rules in a future release as this change is from 0.3.6 to 0.3.7 only.

johnynek commented:
yeah, we can suppress those errors.

They are only for scalajs and they are only if for some weird reason users accessed that BitSet abstraction code, and I don't know why they would. So, I would bet it will impact exactly zero users.

Also, let's add an issue to remove isScalaJs/isScalaJvm when we release a breaking change.

i10416 commented on core/shared/src/test/scala/cats/parse/BitSetTest.scala:
```diff
@@ -24,13 +24,6 @@ package cats.parse
 import org.scalacheck.Prop.forAll
 
 class BitSetTest extends munit.ScalaCheckSuite {
-  test("isScalaJs/isScalaJvm is consistent") {
```
sure

i10416 commented:
@johnynek 
Can I ask you to review this PR? 

Just a reminder. I'm not in a hurry, of course :)

johnynek commented:
Sorry I missed that you had updated this. It looks good to me. If CI is green I'll merge. Thanks so much for working so much to get this done within the binary compatibility constraints.

i10416 commented:
Sooooooorry! I will fix this by tommow! 🙇

johnynek commented:
huh... the CI does not seem to be running.

i10416 commented:
Strang...

johnynek commented:
Thanks for pushing this one through!
